core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java

```diff
@@ -114,28 +113,21 @@ public QueryShardContext(QueryShardContext source) {
         this.types = source.getTypes();
     }
 
-    public void parseFieldMatcher(ParseFieldMatcher parseFieldMatcher) {
-        this.parseContext.parseFieldMatcher(parseFieldMatcher);
-    }
-
+    /**
+     * @return the parseFieldMatcher configured in the index settings
+     */
     public ParseFieldMatcher parseFieldMatcher() {
-        return parseContext.parseFieldMatcher();
+        return indexSettings.getParseFieldMatcher();
```

Comment:
I hadn't noticed this before, but isn't there a circular dependency between IndexSettings and ParseFieldMatcher. I think I would remove the instance within IndexSettings if possible. Doesn't seem hard.